### PR TITLE
chore: publish metadata as gitlab artifact

### PR DESCRIPTION
**Pull request by ggera**
## fixes KILTProtocol/ticket#2241
Generates metadata and publishes json value to s3 bucket
Sample here https://kilt-protocol.org/peregrine/peregrine-metadata.json

## Pending

 - [x] SRTool Docker image that includes the new subwasm 0.19 with support for `json+scale` output format. PR: https://github.com/paritytech/srtool/pull/60

### CODE REVIEW

**Review by ntn-x2 — COMMENTED** [review]
As far as I can see this is definitely good. But I cannot get the SDK to use this properly inside the `augment-api` package. @rflechtner do you have any insights into how we could use this directly instead of pulling the metadata from a running node?

**Comment by ntn-x2** [comment]
Probably scope of a different PR, but subwasm also allows to run diffs between two metadata. We can discuss what is the best strategy to execute those checks, but I would say that running it when opening a PR is merged into `master` might be a good time.

File: .maintain/s3-upload.sh [hunk 1]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
+
+aws configure set region eu-central-1
+
+if [[ "$1" == "spiritnet" ]]; then
+aws s3 cp /tmp/spiritnet-metadata.json s3://$S3_BUCKET/spiritnet/$2/
```
[the remark this answers is not in the document]
Comment (ggera) [reply .maintain/s3-upload.sh]: already added `$1 == spiritnet`

**Comment by ggera** [comment]
> As far as I can see this is definitely good. But I cannot get the SDK to use this properly inside the `augment-api` package. @rflechtner do you have any insights into how we could use this directly instead of pulling the metadata from a running node?

This is expected metadata as json https://kilt-protocol.org/peregrine/peregrine-metadata.json

**Comment by rflechtner** [comment]
> As far as I can see this is definitely good. But I cannot get the SDK to use this properly inside the `augment-api` package. @rflechtner do you have any insights into how we could use this directly instead of pulling the metadata from a running node?

The augment-api script expects the result of the `state_getMetadata` rpc call, which as far as I can see a JSON object with a `result` property whose value is the scale encoded metadata in hex encoding.
I don't know what the best way to scale-encode the JSON example Gera posted is; in theory it should work with
```
import { Metadata, TypeRegistry} from '@polkadot/types';
reg = new TypeRegistry()
meta = new Metadata(reg, json)
meta.toHex()
```
Doesn't really work for me though rn, but that does seem to do the trick:
```
import { TypeRegistry } from '@polkadot/types';
meta = new TypeRegistry().createType('MetadataV14', json.V14)
meta.toHex()
```

**Comment by rflechtner** [comment]
I assume subwasm also has a binary export format for the metadata if you're passing the --json flag? That could be just what we need. Should we try uploading that?

**Comment by ntn-x2** [comment]
I have been playing on the rust server with the latest version of subwasm, and I also get a JSON with a single `V14` key with properties `types`, `pallets`, `extrinsic`, and `ty`. I think the difference could be in how the RPC nodes returns this information. There's also such a thing as a `magicNumber`, which I guess the script expects when querying from the RPC node `state_getMetadata`. We should bypass this and simply parse a metadata document somehow. I will dedicate a bit more time to that this morning, and see if I make any progress on that front.

**Comment by ntn-x2** [comment]
Asked this on StackExchange as well: https://substrate.stackexchange.com/questions/6988/how-to-generate-augmented-types-from-v14-metadata-generated-with-subwasm

**Comment by ntn-x2** [comment]
Another related StackExchange issue: https://github.com/polkadot-js/api/issues/4293

**Comment by trusch** [comment]
https://github.com/chevdor/subwasm/releases/tag/v0.19.0 contains the needed functionality to output all kinds of metadata formats. 

Especially there is `--format=json+scale` which produces the metadata as polkadotjs expects it and `--format=scale` which outputs the format that subxt needs.

**Comment by ntn-x2** [comment]
@trusch so as a quick recap, does `--format=json+scale` produce a file with the metadata? That we can serve directly as an artifact? Can we output multiple things at the same time? Can we also easily check what changed between two metadata versions?

**Comment by trusch** [comment]
@ntn-x2 the format flag just specifies the format, but there is now also a --output flag where you can specify a file.

It's not possible to output two different formats at the same time, but running the command is fast enough to do it two times. 

The diff subcommand was not changed since the last release.

**Comment by ntn-x2** [comment]
@trusch thanks for the clarifications. I would say that we can update this PR to export the new format so that the SDK can use it. We will tackle metadata diffs in a separate ticket.

File: .gitlab-ci.yml [hunk 2]
```diff
@@ -105,9 +107,25 @@ build-wasm-spiritnet:
...
+
+upload:
+  stage: upload
+  dependencies:
+    - build-wasm-spiritnet
+    - build-wasm-peregrine
+  image: amazon/aws-cli:2.9.18
+  script:
+    - aws configure set region $AWS_DEFAULT_REGION
+    - aws s3 cp /out/peregrine-metadata.json s3://$S3_BUCKET/peregrine/${CI_COMMIT_TAG}/
```
Comment (ntn-x2) [inline .gitlab-ci.yml:126]: @ggera as far as I know, `CI_COMMIT_TAG` could be empty if the job was simply triggered by a commit push. Would it make sense to have the same logic we have for Docker image tagging? Or maybe push something like `<branch-commit>` if a regular push and `<tag>` if a tag.

**Comment by ntn-x2** [comment]
I pushed a new change with the new format for subwasm, but we still need to wait for a new SRtool release that includes the new subwasm 0.19.

**Comment by ntn-x2** [comment]
Tracking PR: https://github.com/paritytech/srtool/pull/60

**Comment by ggera** [comment]
> it might be enough if we have them as an artifact. Not sure if we need to upload them to s3. Gitlab also stores them for a while.

Removed the s3 upload and updated srtool since new image is pushed @ntn-x2 @trusch

**Comment by ntn-x2** [comment]
The artifacts only include `*.wasm` outputs. Should also the `*.json` be added as artifacts? So that we have a URL to pull the metadata from.

**Review by ntn-x2 — APPROVED** [review]
Should be good enough! Let's give it a try!

**Pull request merged**